### PR TITLE
Global Cl/Cd SRF Conditioning: two-pass integral force feedback

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Condition the Surface Refinement (SRF) head on predicted global aerodynamic integrals (Cl, Cd) via AdaLN. This creates a two-pass architecture: (1) backbone → coarse surface pressure → numerically integrate Cl/Cd, (2) SRF head receives Cl/Cd conditioning → refined surface pressure. The SRF head becomes aware of the global aerodynamic state, enabling locally consistent corrections that satisfy integral constraints.

**Why this should work:**
- PR #2340 proved that Cl/Cd information improves p_tan by -2.9% — but as a loss term, the integral gradient disrupted shared weights (p_in +2.9%). Here, Cl/Cd enters as an **input signal**, not a loss. No gradient conflict.
- The current SRF operates node-by-node without global awareness. A model that knows "this is a high-lift config (Cl=1.8)" vs "drag-dominated (Cd=0.05)" can make more accurate local corrections for tandem wake pressure.
- AdaLN infrastructure already exists (`--adaln_output`) — this extends it to SRF specifically with aerodynamic conditioning.
- Two-pass architectures are proven: AlphaFold2's iterative refinement, diffusion model conditioning, and hierarchical prediction all use output-conditioned refinement.

### Step 1 — Add flags to Config:
```python
clcd_srf_conditioning: bool = False   # condition SRF head on predicted Cl/Cd via AdaLN
clcd_aux_weight: float = 0.0         # optional auxiliary Cl/Cd supervision weight (0 = no aux loss)
```

### Step 2 — Add Cl/Cd integration helper function:
```python
def integrate_clcd(surface_pressure, surface_normals, arc_lengths, is_surface_mask):
    """Numerically integrate surface pressure to get Cl, Cd.
    
    Args:
        surface_pressure: [B, N, 1] predicted pressure (in Cp space)
        surface_normals: [B, N, 2] unit normals (nx, ny) at surface nodes
        arc_lengths: [B, N, 1] panel arc lengths (ds) at surface nodes
        is_surface_mask: [B, N] boolean surface node mask
    
    Returns:
        clcd: [B, 2] tensor of [Cl, Cd] per sample
    """
    # Cl = -∫ Cp * ny * ds,  Cd = -∫ Cp * nx * ds  (trapezoidal on surface panels)
    p = surface_pressure.squeeze(-1) * is_surface_mask.float()  # [B, N]
    ds = arc_lengths.squeeze(-1) * is_surface_mask.float()      # [B, N]
    nx, ny = surface_normals[:, :, 0], surface_normals[:, :, 1]
    Cl = -(p * ny * ds).sum(dim=1, keepdim=True)  # [B, 1]
    Cd = -(p * nx * ds).sum(dim=1, keepdim=True)   # [B, 1]
    return torch.cat([Cl, Cd], dim=-1)  # [B, 2]
```

**Note:** You'll need to extract surface normals and arc lengths from the data. Check `prepare.py` for available fields. If normals aren't directly available, approximate using finite differences on surface coordinates: `n = normalize(perp(x[i+1] - x[i-1]))`. Arc lengths: `ds[i] = |x[i+1] - x[i-1]| / 2`.

### Step 3 — Add Cl/Cd conditioning MLP to the model:
```python
if self.clcd_srf_conditioning:
    self.clcd_encoder = nn.Sequential(
        nn.Linear(2, 64),
        nn.GELU(),
        nn.Linear(64, 64),
    )
    # AdaLN modulation: scale and shift for each SRF layer
    self.clcd_adaln = nn.Linear(64, surface_refine_hidden * 2)  # scale + shift
```

### Step 4 — Two-pass forward in the SRF head:

**Pass 1:** Run backbone → pressure head → get coarse surface pressure predictions.
**Cl/Cd computation:** Integrate coarse pressure on surface → get [Cl, Cd] per sample.
**Pass 2:** Encode [Cl, Cd] → 64-dim vector → AdaLN scale/shift → apply to SRF hidden layers.

In the SRF forward, after each hidden layer's LayerNorm:
```python
if self.clcd_srf_conditioning:
    clcd_vec = self.clcd_encoder(clcd)          # [B, 64]
    scale, shift = self.clcd_adaln(clcd_vec).chunk(2, dim=-1)  # [B, hidden]
    # Expand to [B, N_surface, hidden] for surface nodes
    h = h * (1 + scale.unsqueeze(1)) + shift.unsqueeze(1)
```

Gradient flows through both passes — the backbone learns that its coarse pressure prediction affects SRF conditioning.

### Step 5 — Optional auxiliary Cl/Cd supervision:
If `--clcd_aux_weight > 0`, add `MSE(pred_Cl, true_Cl) + MSE(pred_Cd, true_Cd)` to the loss. This requires true Cl/Cd in the dataset. If not available, skip this and rely on end-to-end gradient only (set weight to 0).

### Step 6 — Run experiments:
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent fern --wandb_name "fern/clcd-srf-cond-s42" --wandb_group clcd-srf-conditioning \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --clcd_srf_conditioning

CUDA_VISIBLE_DEVICES=1 python train.py \
  --agent fern --wandb_name "fern/clcd-srf-cond-s73" --wandb_group clcd-srf-conditioning \
  --seed 73 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --clcd_srf_conditioning
```

**Important:** Ensure `torch.compile` compatibility — avoid data-dependent shapes in the Cl/Cd integration. Use the surface mask for zeroing instead of indexing to keep tensor shapes static.

## Baseline (PR #2350, 2-seed avg)

| p_in | p_oodc | p_tan | p_re |
|------|--------|-------|------|
| 11.90 | 7.35 | 27.20 | 6.40 |

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature`

## Round 40 Bold Context

This is an **architectural innovation** — a two-pass forward with output-conditioned refinement. PR #2340 proved that Cl/Cd information helps p_tan (-2.9%) but can't be used as a loss without disrupting other metrics. This experiment delivers the same Cl/Cd signal as an **architectural conditioning input** to SRF, avoiding gradient conflicts entirely.

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: **Baseline updated.** PR #2357 (Vortex-Panel Induced Velocity) was just merged. Rebase onto latest `noam` branch and add these flags to your training command:

```
--vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64
```

New baseline metrics to beat (2-seed avg):
| p_in | p_oodc | p_tan | p_re |
|------|--------|-------|------|
| 11.872 | 7.459 | 26.319 | 6.229 |

**Comment by tcapelle** [comment]
STUDENT fern:

Acknowledged baseline update (PR #2357 vortex-panel velocity merged). Rebased onto latest `noam`, resolved merge conflicts, and restarted experiments with the additional `--vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64` flags.

New baseline to beat:
| p_in | p_oodc | p_tan | p_re |
|------|--------|-------|------|
| 11.872 | 7.459 | 26.319 | 6.229 |

Experiments running on GPUs 0-1 (seeds 42 and 73).

**Comment by tcapelle** [comment]
STUDENT fern:

**Bug fix: Added `.detach()` before Cl/Cd integration in training loop.**

The researcher-agent identified a critical issue: without `.detach()` on the coarse pressure before computing Cl/Cd via numerical integration, the SRF loss gradients flow back through the integration path into the backbone. This reintroduces exactly the gradient conflict the two-pass architecture is designed to eliminate.

**Fix applied:** `pred[:, :, 2:3].detach()` at line 2230 (training) and line 2950 (validation, for consistency).

Previous experiments (runs `qv57zfbw` and `nnefll2t`, ~34 epochs each) were killed and restarted with the fix. New runs are `fern/clcd-srf-cond-detach-s42` and `fern/clcd-srf-cond-detach-s73`.

**Known limitation noted:** The `integrate_clcd_from_coords` function sorts surface nodes globally by x-coordinate, which interleaves fore/aft foil nodes at the overlap region in tandem configurations. This produces noisy tangent vectors at the junction. The conditioning signal may be noisier than optimal for tandem foils. Could be improved by splitting by foil identity before sorting — flagging as a potential follow-up.

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Surface MAE (pressure) — the key metric

| Split | Baseline | Seed 42 | Seed 73 | 2-Seed Avg | Δ |
|-------|----------|---------|---------|------------|---|
| **p_in** | **11.872** | 12.586 | 11.862 | **12.224** | **+3.0%** ❌ |
| **p_oodc** | **7.459** | 7.470 | 7.258 | **7.364** | **-1.3%** ✅ |
| **p_tan** | **26.319** | 32.678 | 33.169 | **32.923** | **+25.1%** ❌❌ |
| **p_re** | **6.229** | 6.245 | 6.309 | **6.277** | **+0.8%** ❌ |

### Full Surface MAE

| Split | Metric | Seed 42 | Seed 73 |
|-------|--------|---------|---------|
| in_dist | Ux / Uy / p | 2.794 / 0.768 / 12.586 | 2.804 / 0.992 / 11.862 |
| ood_cond | Ux / Uy / p | 2.044 / 0.566 / 7.470 | 1.710 / 0.903 / 7.258 |
| tandem | Ux / Uy / p | 3.701 / 1.223 / 32.678 | 3.753 / 1.417 / 33.169 |
| ood_re | Ux / Uy / p | 1.649 / 0.410 / 6.245 | 1.535 / 0.703 / 6.309 |

### val_loss

| | Seed 42 | Seed 73 |
|---|---------|---------|
| val_loss | 0.627 | 0.625 |
| best_epoch | 155 | 154 |

### Command

```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=<gpu> python train.py \
  --agent fern --wandb_name "fern/clcd-srf-cond-detach-s<seed>" --wandb_group clcd-srf-conditioning \
  --seed <seed> --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --clcd_srf_conditioning
```

### W&B runs
- Seed 42: `e3a56wk5` (155 epochs, 180-min timeout)
- Seed 73: `d5cn0xux` (154 epochs, 180-min timeout)

### What happened

**Negative result.** The Cl/Cd SRF conditioning via AdaLN two-pass architecture hurt performance, especially on tandem transfer (+25.1%). Only p_oodc showed a marginal improvement (-1.3%).

**Critical bug fixed mid-experiment:** The initial implementation lacked `.detach()` on the coarse pressure before Cl/Cd integration. Without it, SRF loss gradients flowed back through the integration path into the backbone — reintroducing the exact gradient conflict this architecture was designed to avoid. The `.detach()` fix was applied and experiments restarted.

**Why it likely failed:**

1. **Tandem foil interleaving problem.** The `integrate_clcd_from_coords` function sorts all surface nodes globally by x-coordinate. In tandem configurations, fore-foil TE and aft-foil LE overlap in x, causing the sort to interleave nodes from both foils. The resulting finite-difference tangents and normals at the junction are nonsensical (connecting across the gap between foils). This means the Cl/Cd conditioning signal is corrupted for the tandem cases where it matters most.

2. **Noisy conditioning signal.** Even with `.detach()`, the Cl/Cd values derived from coarse predictions are inherently noisy (the model is still learning). The AdaLN modulation injects this noise directly into every LayerNorm, potentially destabilizing SRF head training. The zero-initialized AdaLN should make the initial effect small, but as training progresses and the AdaLN learns non-zero scale/shift, the noise compounds.

3. **Cl/Cd may not carry enough information.** Two scalar integrals (Cl, Cd) are a very lossy summary of the surface pressure distribution. The SRF head already has access to the full coarse prediction via its input features — Cl/Cd conditioning adds very little information that isn't already available.

### Suggested follow-ups

1. **Fix tandem interleaving first.** Split surface nodes by foil identity (using saf_norm threshold) before sorting and integrating. Output 4D [Cl_fore, Cd_fore, Cl_aft, Cd_aft] instead of global 2D. This is a prerequisite for the conditioning signal to be meaningful on tandem.

2. **Richer conditioning signals.** Instead of scalar Cl/Cd, condition on a few Fourier coefficients of the surface pressure distribution, or on the backbone's hidden representation directly (cross-attention from SRF to backbone features). These carry far more information.

3. **This direction may be fundamentally limited.** The SRF head already receives the coarse prediction as input. Scalar force integral conditioning is strictly less information. The hypothesis that "Cl/Cd as input avoids gradient conflict" is correct in principle, but the conditioning signal may be too weak to matter. The gradient conflict issue from PR #2340 might be better addressed through other means (e.g., stop-gradient on the pressure channel fed to auxiliary losses, or a separate Cl/Cd prediction head that doesn't backprop through the pressure channel).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2363 because ADVISOR: Closing — Cl/Cd SRF conditioning is a dead end. p_tan +25.1% is catastrophic.

**Results (2-seed avg):** p_in=12.22 (+3.0%), p_oodc=7.36 (-1.3%), p_tan=32.92 (+25.1%), p_re=6.28 (+0.8%)

**Your analysis is spot-on.** Three fundamental issues:
1. **Tandem interleaving:** Global x-sort interleaves fore/aft nodes, corrupting Cl/Cd for the exact cases that matter most. Excellent catch.
2. **Information bottleneck:** 2 scalars (Cl, Cd) are strictly less information than the full coarse prediction already available to the SRF. The conditioning cannot add meaningful signal.
3. **Noisy conditioning:** Coarse-pass Cl/Cd from a still-learning model injects noise via AdaLN. The zero-init delays but doesn't prevent this.

**On your suggested follow-ups:** The tandem interleaving fix is valid but the fundamental information bottleneck (point 2) makes this direction unlikely to help even if fixed. The SRF already sees the full coarse prediction — conditioning on its 2D integral is provably less informative.

Your insight about 'richer conditioning signals' (Fourier coefficients, cross-attention from SRF to backbone) is closer to what frieren (#2379, ANP decoder) is testing. That's the more promising direction.

New assignment coming shortly. Thank you for the thorough failure analysis — it saves GPU time for the whole team..
